polygeist/lib/Dialect/Polygeist/Transforms/DetectReduction.cpp

```diff
@@ -372,13 +425,25 @@ class LoopReductionIter {
         return WalkResult::interrupt();
       }
 
+      ReductionOp Candidate(Load, CandidateStores[0], OtherLoads);
       Operation *MayAliasOp = nullptr;
       if (Loop.getLoopBody()
               .walk([&](Operation *Op) {
                 if (Op == Load || Op == CandidateStores[0] ||
                     llvm::find(OtherLoads, Op) != OtherLoads.end())
                   return WalkResult::advance();
                 if (hasMayAliasEffects(Load, *Op, aliasAnalysis)) {
+                  if (NumVersion <= DetectReductionVersionLimit) {
+                    Operation *LoadOp = Load;
+                    Optional<ValuePair> VP =
+                        polygeist::VersionConditionBuilder::
+                            getAccessorPairUsedByOperationsDefinedOutsideOfLoop(
+                                Loop, *LoadOp, *Op);
+                    if (VP.has_value()) {
+                      Candidate.addRequireNoOverlapAccessorPairs(*VP);
```

Comment:
Thinking about how to make the design easier to understand. A Candidate could be one that doesn't require versioning. We could also have another type of Candidate that does require versioning (VersioningCandidate). Perhaps make sense for a `VersioningCandidate` to also be a Candidate (so a derived class), perhaps not. 

Only the VersioningCandidate should contain the accessor pairs. While processing Candidates, only VersioningCandidates require versioning the loop so they trigger the code that does the versioning.

Reply:
Will work on a better design in a different PR.